### PR TITLE
More ELMo functionality

### PR DESCRIPTION
**Pull request by matt-peters**
* `Elmo` class
* `ElmoTokenEmbedder`

### CODE REVIEW

File: allennlp/modules/elmo_lstm.py [hunk 1]
```diff
@@ -228,3 +232,61 @@ def _lstm_forward(self,
...
+
+                    # handle the different gate order convention
+                    for torch_w, tf_w in [[input_weights, tf_input_weights],
+                                          [recurrent_weights, tf_recurrent_weights]]:
+                        torch_w[(1 * cell_size):(2 * cell_size), :] = tf_w[(2 * cell_size):(3 * cell_size), :]
+                        torch_w[(2 * cell_size):(3 * cell_size), :] = tf_w[(1 * cell_size):(2 * cell_size), :]
+
+                    lstm.input_linearity.weight.data.copy_(torch.FloatTensor(input_weights))
+                    lstm.state_linearity.weight.data.copy_(torch.FloatTensor(recurrent_weights))
+                    lstm.input_linearity.weight.requires_grad = False
```
Comment (DeNeutoy) [inline allennlp/modules/elmo_lstm.py]: Can you add the fact that these weights are not trainable to the docstring?

File: allennlp/modules/elmo.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        # apply the highway layers (batch_size * sequence_length, n_filters)
+        token_embedding = self._highways(token_embedding)
+
+        # final projection  (batch_size * sequence_length, embedding_dim)
+        token_embedding = self._projection(token_embedding)
+
+        # reshape to (batch_size, sequence_length, embedding_dim)
+        batch_size, sequence_length, _ = character_ids_with_bos_eos.size()
+
+        return {
```
Comment (DeNeutoy) [inline allennlp/modules/elmo.py]: I would mildly prefer that these return tuples, just because 1) There's only 2 arguments and 2) All of the other modules return tuples, so it's a little surprising. I appreciate the reasons, so if you feel strongly, fine.

File: allennlp/modules/elmo.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+        Parameters
+        ----------
+        inputs: ``torch.autograd.Variable``
+            Shape ``(batch_size, timesteps, 50)`` of character ids representing the current batch.
+
+        Returns
+        -------
+        Dict with keys:
+
+        ``'elmo'``: ``List[torch.autograd.Variable]``
```
Comment (DeNeutoy) [inline allennlp/modules/elmo.py]: Nit: can these be called `elmo_representations` or something?

File: allennlp/modules/elmo.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+    In the implementation below, we learn separate scalar weights for each output layer,
+    but only run the biLM once on each input sequence for efficiency.
+
+    Parameters
+    ----------
+    options_file : str
+        ELMo JSON options file
+    weight_file : str
+        ELMo hdf5 weight file
+    num_elmo_layers: int
```
Comment (DeNeutoy) [inline allennlp/modules/elmo.py]: Can we call this `num_scalar_mixture_outputs` or something? I would assume `num_elmo_layers` would mean the number of layers in the model, rather than  what it actually is.

File: allennlp/modules/token_embedders/elmo_token_embedder.py [hunk 5]
```diff
@@ -1,259 +1,60 @@
...
-    Compute context sensitive token representation using pretrained biLM.
+    Compute a single layer of ELMo representations.
 
-    This embedder has input character ids of size (batch_size, sequence_length, 50)
-    and returns (batch_size, sequence_length + 2, embedding_dim), where embedding_dim
-    is specified in the options file (typically 512).
-
-    We add special entries at the beginning and end of each sequence corresponding
-    to <S> and </S>, the beginning and end of sentence tokens.
+    This class servers as a convenience when you only want to use one layer of
```
Comment (DeNeutoy) [inline allennlp/modules/token_embedders/elmo_token_embedder.py]: servers -> serves

File: tests/modules/elmo_test.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                    len(sentence.split()) + 2 for sentence in batch_sentences
+            ]
+            self.assertEqual(lengths.tolist(), expected_lengths)
+
+            # get the expected embeddings and compare!
+            expected_top_layer = [expected_lm_embeddings[k][i] for k in range(3)]
+            for k in range(3):
+                self.assertTrue(
+                        numpy.allclose(
+                                top_layer_embeddings[k, 1:(lengths[k] - 1), :],
```
Comment (DeNeutoy) [inline tests/modules/elmo_test.py]: It would be nice to use the `remove_boundary_tokens` function here so someone doesn't have to think about why you are going from 1:lengths[k] - 1. This crops up a couple of times in the tests, too.

File: tests/modules/elmo_test.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                )
+
+class TestElmo(AllenNlpTestCase):
+    def test_elmo(self):
+        # load the test model
+        options_file = os.path.join(FIXTURES, 'options.json')
+        weight_file = os.path.join(FIXTURES, 'lm_weights.hdf5')
+        elmo = Elmo(options_file, weight_file, 2)
+
+        # Correctness checks are elsewhere, here we just add a shallow test
```
Comment (DeNeutoy) [inline tests/modules/elmo_test.py]: Can you just say where they are (above and in the Simple Tagger model test)

File: allennlp/modules/elmo.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
+import json
+
+from typing import Union, List, Dict, Any
```
Comment (DeNeutoy) [inline allennlp/modules/elmo.py]: Can you group these imports into 3 blocks?

File: allennlp/modules/elmo.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+        with open(cached_path(options_file), 'r') as fin:
+            self._options = json.load(fin)
+        self._weight_file = weight_file
+
+        self.output_dim = self._options['lstm']['projection_dim']
+
+        self._load_weights()
+
+        # Cache the arrays for use in forward -- +1 due to masking.
+        self._beginning_of_sentence_characters = Variable(torch.from_numpy(
```
Comment (DeNeutoy) [inline allennlp/modules/elmo.py]: This seems bad to mutate the id of this here. Why doesn't `ELMoCharacterMapper` take this into account in the first place?

File: allennlp/modules/elmo.py (hunk 4 above)
Comment (matt-peters) [reply allennlp/modules/elmo.py]: I'd like to leave lower level details about `ScalarMix` out of this, and it will make it easier to swap out the weighting if we discover a different a different weighting is better.  How about `num_output_layers`?

File: allennlp/modules/elmo.py (hunk 3 above)
Comment (matt-peters) [reply allennlp/modules/elmo.py]: done

File: allennlp/modules/elmo.py (hunk 8 above)
Comment (matt-peters) [reply allennlp/modules/elmo.py]: done

File: allennlp/modules/elmo.py [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+
+    We add special entries at the beginning and end of each sequence corresponding
+    to <S> and </S>, the beginning and end of sentence tokens.
+
+    Note: this is a lower level class useful for advanced usage.  Most users should
+    use ``ElmoTokenEmbedder`` or ``allennlp.modules.Elmo`` instead.
+
+    Parameters
+    ----------
+    options_file : str
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/modules/elmo.py]: done

File: allennlp/modules/elmo.py [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+    def from_params(cls, params: Params) -> 'Elmo':
+        options_file = params.pop('options_file')
+        weight_file = params.pop('weight_file')
+        num_elmo_layers = params.pop('num_elmo_layers')
+        do_layer_norm = params.pop('do_layer_norm', False)
+        params.assert_empty(cls.__name__)
+        return cls(options_file, weight_file, num_elmo_layers, do_layer_norm)
+
+
+class _ElmoTokenRepresentation(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/modules/elmo.py]: done

File: allennlp/modules/elmo.py [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+                'activation': 'relu',
+                'embedding': {'dim': 4},
+                'filters': [[1, 4], [2, 8], [3, 16], [4, 32], [5, 64]],
+                'max_characters_per_token': 50,
+                'n_characters': 262,
+                'n_highway': 2
+                }
+            }
+
+    something
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/modules/elmo.py]: fixed

File: allennlp/modules/elmo.py (hunk 9 above)
Comment (matt-peters) [reply allennlp/modules/elmo.py]: Agree this breaks encapsulation.  TODO to clean up later.  ELMoCharacterMapper doesn't know anything about masking/padding as that happens in the indexer. So we need to add +1 here.

File: allennlp/modules/elmo.py [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+        Parameters
+        ----------
+        inputs: ``torch.autograd.Variable``
+            Shape ``(batch_size, sequence_length, 50)`` of character ids representing the
+            current batch.
+
+        Returns
+        -------
+        Dict with keys:
+
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/modules/elmo.py]: done

File: allennlp/modules/elmo.py (hunk 2 above)
Comment (matt-peters) [reply allennlp/modules/elmo.py]: I'd rather keep them a dicts, makes changing in backward compatible manner easier.  And the return object is self documenting via the keys.

File: allennlp/modules/elmo_lstm.py (hunk 1 above)
Comment (matt-peters) [reply allennlp/modules/elmo_lstm.py]: done

File: allennlp/modules/encoder_base.py [hunk 14]
```diff
@@ -276,3 +276,10 @@ def _update_states(self,
             # the new states are either of equal size, or smaller, in the case
             # that there are some unused elements (zero-length) for the RNN computation.
             self._states = tuple(new_states)
+
+    @property
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/modules/encoder_base.py]: done

File: allennlp/modules/token_embedders/elmo_token_embedder.py (hunk 5 above)
Comment (matt-peters) [reply allennlp/modules/token_embedders/elmo_token_embedder.py]: done

File: tests/fixtures/elmo/options.json [hunk 15]
```diff
@@ -1 +1 @@
-{"lstm": {"cell_clip": 3, "use_skip_connections": false, "n_layers": 2, "proj_clip": 3, "projection_dim": 16, "dim": 64}, "char_cnn": {"embedding": {"dim": 4}, "filters": [[1, 4], [2, 8], [3, 16], [4, 32], [5, 64]], "n_highway": 2, "n_characters": 262, "max_characters_per_token": 50, "activation": "relu"}}
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply tests/fixtures/elmo/options.json]: done

File: tests/fixtures/elmo/sentences.json [hunk 16]
```diff
@@ -0,0 +1 @@
...oups of people protesting , as in the case of Rongye Adak , a nomad who called for the return of the Dalai Lama and for the freedom of Tibet during the Lithang Horse Racing Festival , in eastern Tibet .", "James Duncan , head of transportation at Bournemouth Borough Council , said : \" Our legal team is reviewing the entitlement of taxis to drop and pick up passengers at bus stops , only for as long as is absolutely necessary to fulfil that function and for no other reason .", "To Mo concerning the food log you kept -- Dr. Buchholz recommends the same thing .", "The CBO estimates that only 23 percent of that would be spent in 2009 and 2010 .", "Even so , Democrats slammed Bush as out of touch .", "An information campaign will be launched later to raise awareness of employment rights and how to enforce them .", "At the gallery the concept is less vague , as Ms. Piper cites specific instances of racial violence , political assassinations and the devastation of Hurricane Katrina .", "There have been some exceptions -- such as Medicare in 1965 .", "The government guidance will be reviewed early next year after a period of public comment .", "It wasn 't the most seaworthy of prizes ."]]
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply tests/fixtures/elmo/sentences.json]: done

File: tests/modules/elmo_test.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+from allennlp.data.iterators import BasicIterator
+from allennlp.modules.elmo import _ElmoBiLm, Elmo, _ElmoTokenRepresentation
+from allennlp.data.fields import TextField
+
+FIXTURES = os.path.join('tests', 'fixtures', 'elmo')
+
+
+class TestElmoBiLm(AllenNlpTestCase):
+    def test_elmo_bilm(self):
+        def _load_sentences_embeddings():
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply tests/modules/elmo_test.py]: done

File: tests/modules/elmo_test.py (hunk 7 above)
Comment (matt-peters) [reply tests/modules/elmo_test.py]: done

File: tests/modules/elmo_test.py (hunk 6 above)
Comment (matt-peters) [reply tests/modules/elmo_test.py]: done

File: allennlp/modules/elmo.py (hunk 4 above)
Comment (matt-peters) [reply allennlp/modules/elmo.py]: I used `num_output_representations`
